sklearn/utils/_array_api.py

```diff
@@ -462,13 +462,15 @@ def _weighted_sum(sample_score, sample_weight, normalize=False, xp=None):
             sample_weight_np = None
         return float(numpy.average(sample_score_np, weights=sample_weight_np))
 
+    # We move to cpu device ahead of time since certain devices may not support
+    # float64, but we want the same precision for all devices and namespaces.
     if not xp.isdtype(sample_score.dtype, "real floating"):
-        # We move to cpu device ahead of time since certain devices may not support
-        # float64, but we want the same precision for all devices and namespaces.
         sample_score = xp.astype(xp.asarray(sample_score, device="cpu"), xp.float64)
```

Comment:
Even when `swapaxes` will be fixed in cupy, this `device="cpu"` move will never ever work on cupy.

Maybe we should never attempt to use `device="cpu"` in our code base and instead use the library specific `_convert_to_numpy` in such code.

But since the problem was already present in `main` prior to this PR, maybe we can defer this to a follow-up PR.

Reply:
Hi @ogrisel, thanks for the review! I finally have time to come back to this PR!

I have noticed that there are some issues on `main` too, I will make another PR to fix main before continuing here. :)

Reply:
Ok, I just had a look at other PRs too. It looks like #27232 and #27904 are already addressing this. Right? should I ignore this and [the below comment ](https://github.com/scikit-learn/scikit-learn/pull/27736/files#r1412133290), or do you think there is still something I can do here? :)

Reply:
Indeed let's wait for the `r2_score` related refactoring in #27904 to be merged before addressing this comment.